Bug in map assignment and access

### Describe the bug

```v
module main

struct Instr {
mut:
	a int
	b int
}

fn main() {
	mut map1 := map[string][]&Instr
	instr := &Instr{a:1,b:2}
	arr := [ instr ]
	map1['Hello'] = arr
	map1['Hello'][0].a = 2 // error?
	println(map1['Hello'][0].a)
}
```


### Expected Behavior

print 2

### Current Behavior

```
$ v run main.v 
==================
/tmp/v_1000/main.5661213509702336810.tmp.c:12465: error: ',' expected (got ")")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

```
v run main.v
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 0bd0942

### Environment details (OS name and version, etc.)

OS: linux, Kali GNU/Linux Rolling
Processor: 8 cpus, 64bit, little endian, 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
CC version: cc (Debian 12.2.0-14) 12.2.0

getwd: /home/ibuki/vas
vmodules: /home/ibuki/.vmodules
vroot: /home/ibuki/v
vexe: /home/ibuki/v/v
vexe mtime: 2023-03-12 23:35:02
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.3 0bd0942

Git version: git version 2.39.1
Git vroot status: weekly.2023.10-41-g0bd09429 (12 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -123,6 +123,7 @@ mut:
 	inside_map_infix          bool // inside map<</+=/-= infix expr
 	inside_assign             bool
 	inside_map_index          bool
+	inside_array_index        bool
 	inside_opt_or_res         bool
 	inside_opt_data           bool
 	inside_if_option          bool

--- a/vlib/v/gen/c/index.v
+++ b/vlib/v/gen/c/index.v
@@ -202,7 +202,14 @@ fn (mut g Gen) index_of_array(node ast.IndexExpr, sym ast.TypeSymbol) {
 				g.write('&')
 			}
 		}
-		g.expr(node.left)
+		if node.left is ast.IndexExpr {
+			g.inside_array_index = true
+			g.expr(node.left)
+			g.inside_array_index = false
+		} else {
+			g.expr(node.left)
+		}
+
 		if node.left_type.has_flag(.shared_f) {
 			if left_is_ptr {
 				g.write('->val')
@@ -393,7 +400,7 @@ fn (mut g Gen) index_of_map(node ast.IndexExpr, sym ast.TypeSymbol) {
 			g.typ(elem_type.clear_flag(.option).clear_flag(.result))
 		}
 	}
-	get_and_set_types := elem_sym.kind in [.struct_, .map]
+	get_and_set_types := elem_sym.kind in [.struct_, .map, .array]
 	if g.is_assign_lhs && !g.is_arraymap_set && !get_and_set_types {
 		if g.assign_op == .assign || info.value_type == ast.string_type {
 			g.is_arraymap_set = true
@@ -434,6 +441,7 @@ fn (mut g Gen) index_of_map(node ast.IndexExpr, sym ast.TypeSymbol) {
 			g.write('${zero} })))')
 		}
 	} else if g.inside_map_postfix || g.inside_map_infix || g.inside_map_index
+		|| g.inside_array_index
 		|| (g.is_assign_lhs && !g.is_arraymap_set && get_and_set_types) {
 		zero := g.type_default(info.value_type)
 		if node.is_setter {

--- a/vlib/v/tests/map_complex_array_test.v
+++ b/vlib/v/tests/map_complex_array_test.v
@@ -1,0 +1,19 @@
+struct Instr {
+mut:
+	a int
+	b int
+}
+
+fn test_map_complex_array() {
+	mut map1 := map[string][]&Instr{}
+	instr := &Instr{
+		a: 1
+		b: 2
+	}
+	arr := [instr]
+	map1['Hello'] = arr
+	map1['Hello'][0].a = 2
+	println(map1['Hello'][0].a)
+	assert map1['Hello'][0].a == 2
+	assert map1['Hello'][0].b == 2
+}

